fragment/fragment/src/main/java/androidx/fragment/app/strictmode/FragmentStrictMode.java

```diff
@@ -0,0 +1,197 @@
+     */
+    public interface OnViolationListener {
+
+        /** Called on a VM policy violation. */
+        void onViolation(@NonNull Violation violation);
+    }
+
+    /**
+     * {@link FragmentStrictMode} policy applied to a certain thread.
+     *
+     * <p>This policy can either be enabled globally using {@link #setDefaultPolicy} or for a
+     * specific {@link FragmentManager} using {@link FragmentManager#setStrictModePolicy(Policy)}.
+     * The current policy can be retrieved using {@link #getDefaultPolicy} and
+     * {@link FragmentManager#getStrictModePolicy} respectively.
+     *
+     * <p>Note that multiple penalties may be provided and they're run in order from least to most
+     * severe (logging before process death, for example). There's currently no mechanism to choose
+     * different penalties for different detected actions.
+     */
+    public static final class Policy {
+        private final Set<Flag> flags;
+        private final OnViolationListener listener;
+
+        /** The default, lax policy which doesn't catch anything. */
+        public static final Policy LAX = new Policy(new HashSet<Flag>(), null);
+
+        private Policy(@NonNull Set<Flag> flags, @Nullable OnViolationListener listener) {
+            this.flags = Collections.unmodifiableSet(flags);
+            this.listener = listener;
+        }
+
+        /**
+         * Creates {@link Policy} instances. Methods whose names start with {@code detect} specify
+         * what problems we should look for. Methods whose names start with {@code penalty} specify
+         * what we should do when we detect a problem.
+         *
+         * <p>You can call as many {@code detect} and {@code penalty} methods as you like. Currently
+         * order is insignificant: all penalties apply to all detected problems.
+         */
+        public static final class Builder {
+            private final Set<Flag> flags;
+            private OnViolationListener listener;
+
+            /** Create a Builder that detects nothing and has no violations. */
+            public Builder() {
+                flags = new HashSet<>();
+            }
+
+            /** Log detected violations to the system log. */
+            @NonNull
+            public Builder penaltyLog() {
+                flags.add(Flag.PENALTY_LOG);
+                return this;
+            }
+
+            /**
+             * Crash the whole process on violation. This penalty runs at the end of all enabled
+             * penalties so you'll still get to see logging or other violations before the process
+             * dies.
+             */
+            @NonNull
+            public Builder penaltyDeath() {
+                flags.add(Flag.PENALTY_DEATH);
+                return this;
+            }
+
+            /** Call #{@link OnViolationListener#onViolation} for every violation. */
+            @NonNull
+            public Builder penaltyListener(@NonNull OnViolationListener listener) {
```

Comment:
`StrictMode` also requires an `Executor` here on which the listener is called. Could possibly also make sense here :thinking:

Reply:
If we can skip it here we should. Ideally this is as simple as possible.

Reply:
I think there's some interesting implications around this, particularly when we want all listeners to be triggered prior to processing the `penaltyDeath` that might inform the structure we want to provide here, even if we don't expose this directly.

For example, do we really want a direct executor to be the default? I could imagine that using the `FragmentHostCallback`'s `Handler` (which, for `FragmentActivity` is the main thread) would be another valid default that would align with most/all of the actual cases we have in mind.

Do you think this work is feasible to do (without a massive rewrite) as a follow up CL? If so, I am okay with just making it more explicit in the Javadoc here with what thread this is being called on (right now, it seems like 'the thread that triggered the policy violation').

One other case I'd like for us to keep in mind is what coroutine friendly API could we build. I don't expect that as part of this initial CL (it is big enough as is), but I'd hate to close that avenue off entirely.

Reply:
Moving it to the `Handler` of the `FragmentHostCallback` sounds like a good idea to me. Should be doable without a big rewrite. For now I added a comment and I'll work on a follow-up PR for this once it is landed.

Reply:
Yeah, I expect that we should do a check similar to what [`ensureExecReady()` does](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:fragment/fragment/src/main/java/androidx/fragment/app/FragmentManager.java;l=1947) to either 1) synchronously trigger listeners / penalty death if we're already on the right looper or 2) post to the `Handler` with two messages: one for dispatching to listeners, the second to trigger the penalty death.

SGTM to do this in a follow up CL, preferably attached to the same underlying issue.